File: JavaScript/0078. Subsets.js
// Given a set of distinct integers, nums, return all possible subsets (the power set).
//
// Note: The solution set must not contain duplicate subsets.
//
// Example:
//
// Input: nums = [1,2,3]
// Output:
//   [
//     [3],
//     [1],
//     [2],
//     [1,2,3],
//     [1,3],
//     [2,3],
//     [1,2],
//     []
//   ]

/**
 * @param {number[]} nums
 * @return {number[][]}
 */

// 1) Backtracking
// Similar
// 46. Permutations
const subsets1 = (nums) => {
  const res = [];

  const go = (cur, rest, start) => {
    res.push(cur);

    for (let i = start; i < rest.length; i++) {
      go(
        [...cur, rest[i]],
        [...rest.slice(0, i), ...rest.slice(i + 1)],
        start,
      );
      start++;
    }
  }

  go([], nums, 0);
  return res;
};

// 2) Similar to 1)
const subsets = (nums) => {
  const res = [];

  const go = (cur, rest, l) => {
    if (l === cur.length) {
      res.push(cur);
      return;
    }

    for (let i = 0; i < rest.length; i++) {
      go(
        [...cur, rest[i]],
        [...rest.slice(i + 1)],
        l,
      );
    }
  };

  for (let i = 0; i <= nums.length; i++) {
    go([], nums, i);
  }
  return res;
};

File: Python/0078. Subsets.py
# Given an integer array nums of unique elements, return all possible subsets (the power set).
# The solution set must not contain duplicate subsets. Return the solution in any order.
#
# Example 1:
#
# Input: nums = [1,2,3]
# Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
#
# Example 2:
#
# Input: nums = [0]
# Output: [[],[0]]
#
# Constraints:
#
# 1 <= nums.length <= 10
# -10 <= nums[i] <= 10
# All the numbers of nums are unique.


# Backtracking
# Similar
# 46. Permutations
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def go(cur, rest, l):
            if l == len(cur):
                res.append(cur)
                return
            for i in range(len(rest)):
                go(cur + [rest[i]], rest[i + 1 :], l)

        for i in range(len(nums) + 1):
            go([], nums, i)
        return res
